src/sink/buffer.rs

```diff
@@ -0,0 +1,65 @@
+use std::collections::VecDeque;
+
+use {Poll, Async};
+use sink::{Sink, StartSend, AsyncSink};
+
+/// Sink for the `Sink::buffer` combinator, which buffers up to some fixed
+/// number of values when the underlying sink is unable to accept them.
+#[must_use = "sinks do nothing unless polled"]
+pub struct Buffer<S: Sink> {
+    sink: S,
+    buf: VecDeque<S::SinkItem>,
+    cap: usize,
+    need_close: bool,
+}
+
+pub fn new<S: Sink>(sink: S, amt: usize) -> Buffer<S> {
+    Buffer {
+        sink: sink,
+        buf: VecDeque::with_capacity(amt),
+        cap: amt,
+        need_close: false,
+    }
+}
+
+impl<S: Sink> Sink for Buffer<S> {
+    type SinkItem = S::SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        try!(self.poll_complete());
```

Comment:
Let's talk on IRC about this -- we've had a bunch of discussions on how this should work with I/O, but seem to still not be on the same page.